### PR TITLE
8344913: Improve -Xlog:cds+map+oop logging for Java mirrors

### PR DESCRIPTION
**Pull request by iklam**
The following information is added to the `-Xlog:cds+map+oop` output for examining the `java/lang/Class` oop instances inside the AOT cache:

- Print the name of the class
- Print the value of the static fields
- Print the  resolved_reference_array for this class, so it's easy to see what constant pool entries (for Lambdas, etc) are resolved.

Below is the new and modified output vs the previous version:

```
- 0x00000000ffe7af28: @@ Object (0xffe7af28) java.lang.Class
+ 0x00000000ffe7af28: @@ Object (0xffe7af28) java.lang.Class Ljava/lang/System;
  0x00000000ffe7af28:   40b7b001 0000035b 001bb638 00000000 001a9b90 00000008 00000000 00000000   ...@[...8.......................
  0x00000000ffe7af48:   00000013 0000000a 00000000 00000000 00000000 00000000 00000000 00000000   ................................
  0x00000000ffe7af68:   00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000   ................................
  0x00000000ffe7af88:   00000000 00000000 00000000 fff04c58 00000000 00000000 00000000 00000000   ............XL..................
  0x00000000ffe7afa8:   00000000 00000000 00000000 00000000 00000000 00000000                     ........................
   - klass: 'java/lang/Class' 0x00000008001bb638
   - fields (19 words):
   - private volatile transient 'classRedefinedCount' 'I' @12  0 (0x00000000)
   - injected 'klass' 'J' @16 0x00000008001a9b90 (marked metadata pointer @0x00000000ffe7af38 )
   - injected 'array_klass' 'J' @24  0 (0x0000000000000000)
   - injected 'oop_size' 'I' @32  19 (0x00000013)
   - injected 'static_oop_field_count' 'I' @36  10 (0x0000000a)
   - private volatile transient 'cachedConstructor' 'Ljava/lang/reflect/Constructor;' @40 null
   - private transient 'name' 'Ljava/lang/String;' @44 null
   - private transient 'module' 'Ljava/lang/Module;' @48 null
   - private final 'classLoader' 'Ljava/lang/ClassLoader;' @52 null
   - private transient 'classData' 'Ljava/lang/Object;' @56 null
   - private transient 'signers' '[Ljava/lang/Object;' @60 null
   - private transient 'packageName' 'Ljava/lang/String;' @64 null
   - private final 'componentType' 'Ljava/lang/Class;' @68 null
   - private volatile transient 'reflectionData' 'Ljava/lang/ref/SoftReference;' @72 null
   - private volatile transient 'genericInfo' 'Lsun/reflect/generics/repository/ClassRepository;' @76 null
   - private volatile transient 'enumConstants' '[Ljava/lang/Object;' @80 null
   - private volatile transient 'enumConstantDirectory' 'Ljava/util/Map;' @84 null
   - private volatile transient 'annotationData' 'Ljava/lang/Class$AnnotationData;' @88 null
   - private volatile transient 'annotationType' 'Lsun/reflect/annotation/AnnotationType;' @92 null
   - transient 'classValueMap' 'Ljava/lang/ClassValue$ClassValueMap;' @96 null
   - injected 'protection_domain' 'Ljava/lang/Object;' @100 null
   - injected 'source_file' 'Ljava/lang/Object;' @104 null
   - injected '<init_lock>' 'Ljava/lang/Object;' @108 0x00000000fff04c58 (0xfff04c58) [I length: 0
+  - signature: Ljava/lang/System;
+  - archived_resolved_references: 0x00000000ffe31b18 (0xffe31b18) [Ljava.lang.Object; length: 29
+ - ---- static fields (10):
+  - public static final 'in' 'Ljava/io/InputStream;' @112 null
+  - public static final 'out' 'Ljava/io/PrintStream;' @116 null
+  - public static final 'err' 'Ljava/io/PrintStream;' @120 null
+  - private static 'initialIn' 'Ljava/io/InputStream;' @124 null
+  - private static 'initialErr' 'Ljava/io/PrintStream;' @128 null
+  - private static 'notSupportedJnuEncoding' 'Ljava/lang/String;' @132 null
+  - private static volatile 'cons' 'Ljava/io/Console;' @136 null
+  - private static 'props' 'Ljava/util/Properties;' @140 null
+  - private static 'lineSeparator' 'Ljava/lang/String;' @144 null
+  - static 'bootLayer' 'Ljava/lang/ModuleLayer;' @148 null
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8344913](https://bugs.openjdk.org/browse/JDK-8344913): Improve -Xlog:cds+map+oop logging for Java mirrors (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22345/head:pull/22345` \
`$ git checkout pull/22345`

Update a local copy of the PR: \
`$ git checkout pull/22345` \
`$ git pull https://git.openjdk.org/jdk.git pull/22345/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22345`

View PR using the GUI difftool: \
`$ git pr show -t 22345`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22345.diff">https://git.openjdk.org/jdk/pull/22345.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22345#issuecomment-2495725993)
</details>

### CODE REVIEW

File: src/hotspot/share/cds/archiveBuilder.cpp [hunk 1]
```diff
@@ -1412,7 +1458,7 @@ class ArchiveBuilder::CDSMapLogger : AllStatic {
   // the narrowOop version of the requested address.
   //     0x00000007ffc7e840 (0xfff8fd08) java.lang.Class
   //     0x00000007ffc000f8 (0xfff8001f) [B length: 11
-  static void print_oop_with_requested_addr_cr(outputStream* st, oop source_oop, bool print_addr = true) {
+  static void print_oop_with_requested_addr_cr(outputStream* st, oop source_oop, bool print_addr = true, bool print_aot_init = false) {
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/archiveBuilder.cpp:1461]: Pre-existing but it seems very odd to me to have a function called `print_oop_with_requested_addr` and then have a parameter to control whether we print the addr.

File: src/hotspot/share/cds/archiveBuilder.cpp [hunk 2]
```diff
@@ -1428,7 +1474,29 @@ class ArchiveBuilder::CDSMapLogger : AllStatic {
...
+            Klass* src_klass = java_lang_Class::as_Klass(scratch_mirror);
+            if (src_klass != nullptr && src_klass->is_instance_klass()) {
+              InstanceKlass* buffered_klass =
+                ArchiveBuilder::current()->get_buffered_addr(InstanceKlass::cast(src_klass));
+              if (buffered_klass->has_aot_initialized_mirror()) {
+                st->print(" (aot-inited)");
+              }
+            }
+          }
+        }
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/archiveBuilder.cpp:1498]: Is there a reason we wouldn't always print whether or not it is aot-inited? It seems useful information.

**Review by dholmes-ora — COMMENTED** [review]
Seems reasonable, albeit a bit complicated to follow what exactly is being printed where and why.

File: src/hotspot/share/cds/archiveBuilder.cpp [hunk 3]
```diff
@@ -1294,7 +1294,7 @@ class ArchiveBuilder::CDSMapLogger : AllStatic {
 
       if (source_oop != nullptr) {
         // This is a regular oop that got archived.
-        print_oop_with_requested_addr_cr(&st, source_oop, false);
+        print_oop_with_requested_addr_cr(&st, source_oop, false, true);
```
Comment (calvinccheung) [inline src/hotspot/share/cds/archiveBuilder.cpp:1297]: Now that the function takes two boolean args. I think it is good to add comments on those boolean args.

File: src/hotspot/share/cds/archiveBuilder.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/cds/archiveBuilder.cpp]: I changed the function name to `print_oop_info_cr()` and added comments about whether address(es) are printed, and in what format.

File: src/hotspot/share/cds/archiveBuilder.cpp (hunk 3 above)
Comment (iklam) [reply src/hotspot/share/cds/archiveBuilder.cpp]: I added the comment about what the first `false` parameter means. I also removed the second parameter as suggested by David.

**Review by calvinccheung — APPROVED** [review]
Looks good.

**Review by dholmes-ora — APPROVED** [review]
Thanks for the update - looks good.

**Comment by iklam** [comment]
Thanks @dholmes-ora @calvinccheung for the review
/integrate

**Pull request closed without merging**